server/src/main/java/org/elasticsearch/search/fetch/subphase/FetchSourceContext.java

```diff
@@ -85,18 +145,70 @@ public String[] excludes() {
         return this.excludes;
     }
 
+    public Boolean includeVectors() {
+        return this.includeVectors;
+    }
+
+    public boolean filterVectorFields() {
+        return this.includeVectors == null || this.includeVectors == false;
+    }
+
     public boolean hasFilter() {
         return this.includes.length > 0 || this.excludes.length > 0;
     }
 
     public SourceFilter filter() {
-        return new SourceFilter(includes, excludes);
+        return filter(null);
+    }
+
+    public SourceFilter filter(MappingLookup mappingLookup) {
+        if (filterVectorFields() == false) {
+            return new SourceFilter(includes, excludes);
+        } else {
+            if (mappingLookup == null) {
+                throw new IllegalStateException("mappingLookup must not be null when filtering vectors");
+            }
+
+            String[] excludeFields = excludes();
+            List<String> inferenceFieldNames = new ArrayList<>(mappingLookup.inferenceFields().size());
+            for (String inferenceField : mappingLookup.inferenceFields().keySet()) {
+                Mapper mapper = mappingLookup.getMapper(inferenceField);
+                if (mapper instanceof InferenceFieldMapper == false) {
+                    throw new IllegalStateException("Mapper for field [" + inferenceField + "] is not an inference field mapper");
+                }
+
+                InferenceFieldMapper inferenceFieldMapper = (InferenceFieldMapper) mapper;
+                inferenceFieldNames.add(inferenceFieldMapper.getInferenceFieldName());
+            }
+            excludeFields = ArrayUtils.concat(excludeFields, inferenceFieldNames.toArray(new String[0]));
+
+            if (includeVectors != null) {
+                String[] denseVectors = mappingLookup.getFullNameToFieldType()
+                    .entrySet()
+                    .stream()
+                    .filter(entry -> entry.getValue() instanceof DenseVectorFieldMapper.DenseVectorFieldType)
+                    .map(Map.Entry::getKey)
+                    .toArray(String[]::new);
+                excludeFields = ArrayUtils.concat(excludeFields, denseVectors);
+
+                String[] sparseVectors = mappingLookup.getFullNameToFieldType()
+                    .entrySet()
+                    .stream()
+                    .filter(entry -> entry.getValue() instanceof SparseVectorFieldMapper.SparseVectorFieldType)
+                    .map(Map.Entry::getKey)
+                    .toArray(String[]::new);
+                excludeFields = ArrayUtils.concat(excludeFields, sparseVectors);
+            }
+
+            return new SourceFilter(includes, excludeFields);
+        }
     }
 
     public static FetchSourceContext parseFromRestRequest(RestRequest request) {
```

Comment:
We should update the specification of the APIs that use `parseFromRestRequest` to generate a `FetchSourceContext` to add the `_source_include_vectors` param. However, given how large this PR is already, I propose that we do this in a follow-up PR.

I updated the specs for the `get`, `bulk`, and `update` APIs in this PR because we need to use `_source_include_vectors` with those APIs in YAML tests.